Libs/libMacGitverCore/Widgets/ProgressDlg.cpp

```diff
@@ -0,0 +1,261 @@
+    mUpdater.start(250);
+}
+
+ProgressDlg::~ProgressDlg()
+{
+    delete ui;
+}
+
+int ProgressDlg::updateInterval() const
+{
+    return mUpdater.interval();
+}
+
+void ProgressDlg::setUpdateInterval(int msec)
+{
+    if (mUpdater.isActive()) {
+        // restart timer with the new interval
+        mUpdater.start(msec);
+    }
+    else {
+        mUpdater.setInterval(msec);
+    }
+}
+
+void ProgressDlg::addActivity(const QString& description, QObject* activity,
+                              const StepInfo::List& steps)
+{
+    Q_ASSERT(activity);
+
+    Private::ProgressWdgt* a = new Private::ProgressWdgt(description);
+    mActivities[activity] = a;
+
+    QTreeWidgetItem* activityItem(new QTreeWidgetItem);
+    ui->treeProgress->addTopLevelItem(activityItem);
+    ui->treeProgress->setItemWidget(activityItem, 0, a);
+
+    foreach (const StepInfo& sir, steps) {
+        Private::ProgressWdgt* s = new Private::ProgressWdgt(sir.desc);
+        a->mSteps[sir.name] = s;
+
+        QTreeWidgetItem* stepItem(new QTreeWidgetItem);
+        activityItem->addChild(stepItem);
+        ui->treeProgress->setItemWidget(stepItem, 0, s);
+    }
+
+    a->txtStatusInfo->setText(tr("Running..."));
+}
+
+void ProgressDlg::setStatusInfo(QObject* activity, const QString& step,
+                                const QString& text)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->txtStatusInfo->setText(text);
+}
+
+void ProgressDlg::setPercentage(QObject* activity, const QString& step,
+                                qreal percent)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->mPercentage = qMin( qMax(percent, 0.), 1. ) * 100.;
+}
+
+void ProgressDlg::closeEvent( QCloseEvent* ev )
+{
+    if( !mDone )
+    {
+        ev->ignore();
+        return;
+    }
+
+    QDialog::closeEvent( ev );
+}
+
+void ProgressDlg::finished(QObject* activity)
+{
+    Private::ProgressWdgt* a = mActivities[activity];
+    Q_ASSERT(a);
+
+    a->mActive = false;
+    a->txtStatusInfo->setText(tr("Finished!"));
+    a->resultChanged(Private::ProgressWdgt::Result::Ok);
+}
+
+void ProgressDlg::finished(QObject* activity, const QString& step)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->mActive = false;
+    s->mPercentage = 100.;
+    s->resultChanged(Private::ProgressWdgt::Result::Ok);
+}
+
+void ProgressDlg::setError(QObject* activity, const QString& message)
+{
+    Private::ProgressWdgt* a = mActivities[activity];
+    Q_ASSERT(a);
+    a->mActive = false;
+    a->txtStatusInfo->setText(tr("Failed!"));
+    a->setToolTip(tr("<b>Activity failed!</b><hr/>%1").arg(message));
+    a->resultChanged(Private::ProgressWdgt::Result::Error);
+}
+
+void ProgressDlg::updateActivities()
+{
+    Activities::Iterator it = mActivities.begin();
+    while ( it != mActivities.end() ) {
+        Private::ProgressWdgt* a = *it;
```

Comment:
I think that that assumption is bad. The `ProgressDlg` ought to live forever. Hence, when the activity is finished, the progress widgets ought to be removed. However, we should not do that immediately - So maybe this needs a "dismiss" mechanism; and when doing that we might as well introduce a "cancel" mechanism (The UI ought to be the same, but the implementation will be quite different / difficult).

Anyway, if your assumption that a widget is not deleted before the dialog is true, then there's no need to store them inside `QPointer<>`'s.